### ISSUE TITLE
Pleanse help me

### ISSUE THREAD

**Issue by limitnum** [lofrienger/S2ME#8]
“I loaded the best weights, but I got a very bad effect. Could you tell me what may cause it, or supply a new weight for me to try again?”

**Comment by limitnum** [comment]
I have solved this problem , but I don't know why can't start model.eval() mode otherwise emerged above phenomenon

**Comment by lofrienger** [comment]
Hi, model.eval() should work as we do regular validation during training. 
Perhaps you can refer to that part.

**Comment by limitnum** [comment]
> Hi, model.eval() should work as we do regular validation during training. Perhaps you can refer to that part.

thanks for your reply , but I used the weights you supplyed to test , so I start the model.eval() , then got a very bad effect  . When I closed it  , it dose work . So should I use the mode of eval() during testing

**Comment by lofrienger** [comment]
Hi, yes, you should use eval mode during testing. 

Our training script is hybrid, including train, val, and test phases. During the test, we use the saved best_model1, which is a deepcopy of the model in eval mode during the val phase, so we do not don't have to set best_model1.eval() coz it is in the eval model already.

But if you do this in a split way, where you load our provided model and do the test, you should explicitly set model.eval().

Hope the explanation helps.

**Comment by limitnum** [comment]
> Hi, yes, you should use eval mode during testing.
> 
> Our training script is hybrid, including train, val, and test phases. During the test, we use the saved best_model1, which is a deepcopy of the model in eval mode during the val phase, so we do not don't have to set best_model1.eval() coz it is in the eval model already.
> 
> But if you do this in a split way, where you load our provided model and do the test, you should explicitly set model.eval().
> 
> Hope the explanation helps.
Thank you very much for your thoughtful reply. But I'm the latter case(split way). When I set eval(), the model performs very poorly. On the contrary, it performs normally without explicitly setting eval().

**Comment by lofrienger** [comment]
Hi, our test results are logged along with the training/validation process. As you can see in the provided [training log](https://github.com/lofrienger/S2ME/blob/main/model/log.txt).
And I think there should be no deviation between these two test methods. Please check your test code or paste here...

**Comment by limitnum** [comment]
> Hi, our test results are logged along with the training/validation process. As you can see in the provided [training log](https://github.com/lofrienger/S2ME/blob/main/model/log.txt). And I think there should be no deviation between these two test methods. Please check your test code or paste here...

import argparse
import logging
import os
import random
import sys
import copy
from data.data.dataloader import *
from lib.utils import AverageMeter
from lib.utils.metrics import  *
import numpy as np
import torch
import torch.backends.cudnn as cudnn
import torch.optim as optim

from dataloaders.dataset_polyp import (PolypDataset, trsf_train_image_224,
                                       trsf_valid_image_224)
from networks.net_factory import net_factory
from torch.nn.modules.loss import CrossEntropyLoss
from torch.utils.data import DataLoader
from torchvision.transforms import ToPILImage
from tqdm import tqdm
from utils import losses, ramps
from valid import test_polyp_batch
from lib.utils.util import *

def test(model, test_loader, _data_name, res_dir):
    dice_meter = AverageMeter('DICE', ':.4e')
    iou_meter = AverageMeter('IOU', ':.4e')
    # model.eval()
    save_path = f"{res_dir}/{_data_name}/"
    if not os.path.exists(save_path):
        os.makedirs(save_path)

    for i in range(test_loader.size):
        image, gt, name = test_loader.load_data()
        image = image.cuda()
        gt = gt.cuda()
        c_logits= model(image)
        # print(c_logits.size()  , gt.size())
        target_size = (gt.shape[-2], gt.shape[-1])

        c_logits = F.upsample(c_logits, size=target_size, mode='bilinear',
                              align_corners=False)  # torch.Size([1, 2, 500, 574])
        c_logits = torch.softmax(c_logits, dim=1)

        res = torch.argmax(c_logits, dim=1)
        # print(res.size())
        # print(f"gt : {np.unique(gt.cpu().numpy())}   res : {np.unique(res.cpu().numpy())}")
        res_np = res.squeeze().data.cpu().numpy().astype('uint8')
        img = Image.fromarray(res_np * 255 , mode='L')
        img.save(save_path + name)
        # misc.imsave(save_path + name, res.squeeze().data.cpu().numpy())

        dice, iou = DICE_IOU(c_logits, gt)
        dice_meter.update(dice.item(), image.size(0))
        iou_meter.update(iou.item(), image.size(0))

    print(_data_name, ':====> TEST DICE: {:.4f}'.format(dice_meter.avg))
    print(_data_name, ':====> TEST IOU: {:.4f}'.format(iou_meter.avg))

    return dice_meter.avg, iou_meter.avg



if __name__ == "__main__":
    model = net_factory(net_type="unet", in_chns=3, class_num=2)
    model_path_name = os.path.join(r'../model', 'best_model.pth')
    if os.path.isfile(model_path_name):
        print("=> Loading checkpoint '{}'".format(model_path_name))
        checkpoint = torch.load(model_path_name)
        model.load_state_dict(checkpoint)
        model = model.cuda()
        # model.eval()
        # print("=> Loaded saved the best model at (epoch {})".format(checkpoint['epoch']))
    else:
        print("=> No checkpoint found at '{}'".format(model_path_name))
    # print(model)
    res_dic = {}
    # res_dic["args"] = args
    res_dir = './result'
    res_file = f"{res_dir}/test_performance_all_seeds.csv"
    # res_dic["best_epoch"] = checkpoint['epoch']

    for _data_name in ['CVC-300', 'CVC-ClinicDB', 'Kvasir', 'CVC-ColonDB', 'ETIS-LaribPolypDB']:
        data_path = f"../data/data/polyp/TestDB/TestDB/{_data_name}"
        image_root = '{}/images/'.format(data_path)
        gt_root = '{}/masks/'.format(data_path)
        test_loader = polyp_test_dataset(image_root, gt_root)
        dice, iou = test(model, test_loader, _data_name, res_dir)
        res_dic[f"{_data_name}_dice"] = dice
        res_dic[f"{_data_name}_iou"] = iou
    write_add_csv(res_dic, res_file)

**### Here's my code. Could you please help me figure out what's wrong with it, if you have the time? I'd really appreciate it!**

**Comment by lofrienger** [comment]
Hi, you are using different inference and metric computation codes from ours, and I think that should be the reason. 
I recommend that you use ours from valid.py, following our train.py

**Comment by limitnum** [comment]
Because I saved the result（inference is same ， use the “unet” model） as a visualization, it displays poorly when I use eval(), but it works fine when I don't use eval(). This issue is really strange, but it doesn't matter.
Thank you very much for your patient replies.



---Original---
From: "An ***@***.***&gt;
Date: Tue, Jul 22, 2025 00:39 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [lofrienger/S2ME] Pleanse help me (Issue #8)


lofrienger left a comment (lofrienger/S2ME#8)
 
Hi, you are using different inference and metric computation codes from ours, and I think that should be the reason.
 I recommend that you use ours from valid.py, following our train.py
